Error on color box click

## Description

```text
Traceback (most recent call last):
  File "...\sublime3\sublime_plugin.py", line 812, in run_
    return self.run(edit, **args)
  File "color_helper in ...\sublime3\Data\Installed Packages\ColorHelper.sublime-package", line 983, in run
  File "color_helper in ...\sublime3\Data\Installed Packages\ColorHelper.sublime-package", line 891, in show_color_info
  File "color_helper in ...\sublime3\Data\Installed Packages\ColorHelper.sublime-package", line 553, in format_info
  File "...\sublime3\Data\Packages\mdpopups\st3\mdpopups\__init__.py", line 458, in color_box
    border_size, check_size, max_colors, alpha, border_map
  File "...\sublime3\Data\Packages\mdpopups\st3\mdpopups\colorbox.py", line 250, in color_box
    base64.b64encode(color_box_raw(*args, **kwargs)).decode('ascii')
  File "...\sublime3\Data\Packages\mdpopups\st3\mdpopups\colorbox.py", line 97, in color_box_raw
    assert height - (border_size * 2) >= 0, "Border size too big!"
AssertionError: Border size too big!
```

## Support Info

* Sublime Text Build 3152 (64bit, portable)
* Win10 Pro x64 Version 1709
* default `"dpi_scale": 1.0`, but HighDPI screen
* ColorHelper v2.5.1
* Install via PC: True
* mdpopups ver.: 3.0.2
* markdown ver.: 2.6.9
* pygments ver.: 2.1a0
* jinja2 ver.: 2.8

## Steps to Reproduce Issue

1. Open a Color Scheme (`*.tmTheme`) file.
2. Click on a color box.

--- a/color_helper.py
+++ b/color_helper.py
@@ -926,12 +926,9 @@ class ColorHelperCommand(sublime_plugin.TextCommand):
         """Get sizes."""
 
         self.graphic_size = qualify_settings(ch_settings, 'graphic_size', 'medium')
-        try:
-            self.graphic_scale = qualify_settings(ch_settings, 'graphic_scale', 1)
-            if not isinstance(self.graphic_scale, (int, float)) or self.graphic_scale < 0:
-                self.graphic_scale = 1
-        except Exception:
-            self.graphic_scale = 1
+        self.graphic_scale = qualify_settings(ch_settings, 'graphic_scale', None)
+        if not isinstance(self.graphic_scale, (int, float)):
+            self.graphic_scale = None
         top_pad = self.view.settings().get('line_padding_top', 0)
         bottom_pad = self.view.settings().get('line_padding_bottom', 0)
         # Sometimes we strangely get None
@@ -940,13 +937,16 @@ class ColorHelperCommand(sublime_plugin.TextCommand):
         if bottom_pad is None:
             bottom_pad = 0
         box_height = util.get_line_height(self.view) - int(top_pad + bottom_pad) - 6
-        if box_height < 8:
-            box_height = 8
-        box_height = int(box_height * self.graphic_scale)
+        if self.graphic_scale is not None:
+            box_height = box_height * self.graphic_scale
+            self.graphic_size = "small"
+        small = max(box_height, 8)
+        medium = max(box_height * 1.5, 8)
+        large = max(box_height * 2, 8)
         sizes = {
-            "small": (box_height, box_height, box_height * 2),
-            "medium": (int(box_height * 1.5), int(box_height * 1.5), box_height * 2),
-            "large": (int(box_height * 2), int(box_height * 2), box_height * 2)
+            "small": (int(small), int(small), int(small) * 2),
+            "medium": (int(medium), int(medium), int(small) * 2),
+            "large": (int(large), int(large), int(small) * 2)
         }
         self.color_h, self.color_w, self.palette_w = sizes.get(
             self.graphic_size,

--- a/color_helper.py
+++ b/color_helper.py
@@ -926,6 +926,12 @@ class ColorHelperCommand(sublime_plugin.TextCommand):
         """Get sizes."""
 
         self.graphic_size = qualify_settings(ch_settings, 'graphic_size', 'medium')
+        try:
+            self.graphic_scale = qualify_settings(ch_settings, 'graphic_scale', 1)
+            if not isinstance(self.graphic_scale, (int, float)) or self.graphic_scale < 0:
+                self.graphic_scale = 1
+        except Exception:
+            self.graphic_scale = 1
         top_pad = self.view.settings().get('line_padding_top', 0)
         bottom_pad = self.view.settings().get('line_padding_bottom', 0)
         # Sometimes we strangely get None
@@ -936,6 +942,7 @@ class ColorHelperCommand(sublime_plugin.TextCommand):
         box_height = util.get_line_height(self.view) - int(top_pad + bottom_pad) - 6
         if box_height < 8:
             box_height = 8
+        box_height = int(box_height * self.graphic_scale)
         sizes = {
             "small": (box_height, box_height, box_height * 2),
             "medium": (int(box_height * 1.5), int(box_height * 1.5), box_height * 2),

--- a/color_helper.py
+++ b/color_helper.py
@@ -934,6 +934,8 @@ class ColorHelperCommand(sublime_plugin.TextCommand):
         if bottom_pad is None:
             bottom_pad = 0
         box_height = util.get_line_height(self.view) - int(top_pad + bottom_pad) - 6
+        if box_height < 8:
+            box_height = 8
         sizes = {
             "small": (box_height, box_height, box_height * 2),
             "medium": (int(box_height * 1.5), int(box_height * 1.5), box_height * 2),

--- a/color_helper.sublime-settings
+++ b/color_helper.sublime-settings
@@ -43,12 +43,15 @@
     // package is installed.
     "enable_color_picker": true,
 
-    // If the color picker is too big, try playing with this.
-    // Graphics in tooltips usually look better bigger (especially in Hidpi),
-    // but that can make the tooltips really big. If they are too big,
-    // you can play with this setting.  We compromise with medium.
+    // If the color picker is too big or too small, try playing with this.
+    // This is a coarse control which scales image size relative to the
+    // approximate line height. small (1x), medium (1.5x), and large (2x).
     // (small | medium | large)
     "graphic_size": "medium",
+
+    // Fine scaling of image sizes. Overrides `graphic_size`.
+    // Use an integer or floating point value. `null` disables fine scaling.
+    "graphic_scale": null,
 
     // Use https://github.com/weslly/ColorPicker for the color picker if installed.
     "use_color_picker_package": false,

--- a/color_helper_picker.py
+++ b/color_helper_picker.py
@@ -492,6 +492,8 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
         if bottom_pad is None:
             bottom_pad = 0
         self.box_height = self.line_height - int(top_pad + bottom_pad) - 6
+        if self.box_height < 8:
+            self.box_height = 8
         sizes = {
             "small": (int(self.box_height * .85), int(self.box_height * .85), int(self.box_height * 1.0)),
             "medium": (int(self.box_height), int(self.box_height), int(self.box_height * 1.25)),

--- a/color_helper_picker.py
+++ b/color_helper_picker.py
@@ -283,7 +283,7 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
                 '[%s](%s) %s<br>' % (
                     mdpopups.color_box(
                         [color], self.default_border,
-                        border_size=BORDER_SIZE, height=self.box_height, width=self.box_height * 8,
+                        border_size=BORDER_SIZE, height=self.height, width=self.box_height * 8,
                         check_size=check_size
                     ),
                     color,
@@ -340,7 +340,7 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
                 '[%s](%s) %s<br>' % (
                     mdpopups.color_box(
                         [color], self.default_border,
-                        border_size=BORDER_SIZE, height=self.box_height, width=self.box_height * 8,
+                        border_size=BORDER_SIZE, height=self.height, width=self.box_height * 8,
                         check_size=check_size
                     ),
                     color,
@@ -490,12 +490,9 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
 
         settings = sublime.load_settings('color_helper.sublime-settings')
         self.graphic_size = qualify_settings(settings, 'graphic_size', 'medium')
-        try:
-            self.graphic_scale = qualify_settings(settings, 'graphic_scale', 1)
-            if not isinstance(self.graphic_scale, (int, float)) or self.graphic_scale < 0:
-                self.graphic_scale = 1
-        except Exception:
-            self.graphic_scale = 1
+        self.graphic_scale = qualify_settings(settings, 'graphic_scale', None)
+        if not isinstance(self.graphic_scale, (int, float)):
+                self.graphic_scale = None
         self.line_height = util.get_line_height(self.view)
         top_pad = self.view.settings().get('line_padding_top', 0)
         bottom_pad = self.view.settings().get('line_padding_bottom', 0)
@@ -504,22 +501,18 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
             top_pad = 0
         if bottom_pad is None:
             bottom_pad = 0
-        self.box_height = self.line_height - int(top_pad + bottom_pad) - 6
-        if self.box_height < 8:
-            self.box_height = 8
-        self.box_height = int(self.box_height * self.graphic_scale)
+        box_height = self.line_height - int(top_pad + bottom_pad) - 6
+        if self.graphic_scale is not None:
+            box_height = box_height * self.graphic_scale
+            self.graphic_size = "small"
+        small = max(box_height, 8)
+        medium = max(box_height * 1.5, 8)
+        large = max(box_height * 2, 8)
+        self.box_height = int(small)
         sizes = {
-            "small": (int(self.box_height), int(self.box_height), int(self.box_height + self.box_height / 4)),
-            "medium": (
-                int(self.box_height * 1.5),
-                int(self.box_height * 1.5),
-                int(self.box_height * 1.5 + (self.box_height * 1.5) / 4)
-            ),
-            "large": (
-                int(self.box_height * 2),
-                int(self.box_height * 2),
-                int(self.box_height * 2 + (self.box_height * 2) / 4)
-            )
+            "small": (int(small), int(small), int(small + small / 4)),
+            "medium": (int(medium), int(medium), int(medium + medium / 4)),
+            "large": (int(large), int(large), int(large + large / 4))
         }
         self.height, self.width, self.height_big = sizes.get(
             self.graphic_size,

--- a/color_helper_picker.py
+++ b/color_helper_picker.py
@@ -492,7 +492,7 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
         self.graphic_size = qualify_settings(settings, 'graphic_size', 'medium')
         self.graphic_scale = qualify_settings(settings, 'graphic_scale', None)
         if not isinstance(self.graphic_scale, (int, float)):
-                self.graphic_scale = None
+            self.graphic_scale = None
         self.line_height = util.get_line_height(self.view)
         top_pad = self.view.settings().get('line_padding_top', 0)
         bottom_pad = self.view.settings().get('line_padding_bottom', 0)

--- a/color_helper_picker.py
+++ b/color_helper_picker.py
@@ -35,6 +35,7 @@ color_map_size = False
 color_map_style = None
 line_height = None
 default_border = None
+color_scale = None
 
 SPACER = '#00000000'
 
@@ -52,15 +53,18 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
         global color_map_style
         global line_height
         global default_border
+        global color_scale
 
         if (
             color_map is None or
             self.graphic_size != color_map_size or
+            self.graphic_scale != color_scale or
             self.line_height != line_height or
             self.default_border != default_border or
             color_map_style != "square"
         ):
             color_map_size = self.graphic_size
+            color_scale = self.graphic_scale
             color_map_style = "square"
             line_height = self.line_height
             default_border = self.default_border
@@ -167,15 +171,18 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
         global color_map_style
         global line_height
         global default_border
+        global color_scale
 
         if (
             color_map is None or
             self.graphic_size != color_map_size or
+            self.graphic_scale != color_scale or
             self.line_height != line_height or
             self.default_border != default_border or
             color_map_style != "hex"
         ):
             color_map_size = self.graphic_size
+            color_scale = self.graphic_scale
             line_height = self.line_height
             default_border = self.default_border
             color_map_style = "hex"
@@ -483,6 +490,12 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
 
         settings = sublime.load_settings('color_helper.sublime-settings')
         self.graphic_size = qualify_settings(settings, 'graphic_size', 'medium')
+        try:
+            self.graphic_scale = qualify_settings(settings, 'graphic_scale', 1)
+            if not isinstance(self.graphic_scale, (int, float)) or self.graphic_scale < 0:
+                self.graphic_scale = 1
+        except Exception:
+            self.graphic_scale = 1
         self.line_height = util.get_line_height(self.view)
         top_pad = self.view.settings().get('line_padding_top', 0)
         bottom_pad = self.view.settings().get('line_padding_bottom', 0)
@@ -494,10 +507,19 @@ class ColorHelperPickerCommand(sublime_plugin.TextCommand):
         self.box_height = self.line_height - int(top_pad + bottom_pad) - 6
         if self.box_height < 8:
             self.box_height = 8
+        self.box_height = int(self.box_height * self.graphic_scale)
         sizes = {
-            "small": (int(self.box_height * .85), int(self.box_height * .85), int(self.box_height * 1.0)),
-            "medium": (int(self.box_height), int(self.box_height), int(self.box_height * 1.25)),
-            "large": (int(self.box_height * 1.15), int(self.box_height * 1.15), int(self.box_height * 1.35))
+            "small": (int(self.box_height), int(self.box_height), int(self.box_height + self.box_height / 4)),
+            "medium": (
+                int(self.box_height * 1.5),
+                int(self.box_height * 1.5),
+                int(self.box_height * 1.5 + (self.box_height * 1.5) / 4)
+            ),
+            "large": (
+                int(self.box_height * 2),
+                int(self.box_height * 2),
+                int(self.box_height * 2 + (self.box_height * 2) / 4)
+            )
         }
         self.height, self.width, self.height_big = sizes.get(
             self.graphic_size,

--- a/docs/src/markdown/usage.md
+++ b/docs/src/markdown/usage.md
@@ -148,18 +148,27 @@ This will make the color preview box in the [Color Info Panel](#color-info-panel
 
 ### `graphic_size`
 
-Controls the size of generated graphics.  Graphics in the tooltips look best large as Sublime slightly distorts images, but on small screens, some of the tooltips (especially the internal color picker) may be too large.  `graphic_size` can be used to control the size of these generated images.  Valid settings are `small`, `medium`, and `large` where `medium` is the default.
+Coarse scaling of the size of generated graphics.  Graphics in the tooltips look best large as Sublime slightly distorts images, but on different screens, some of the tooltips (especially the internal color picker) may be too large or too small. `graphic_size` can be used to control the size of these generated images.  Scaling is based off line height, and valid settings are `small` (1X), `medium` (1.5X), and `large` (2X). `medium` is the default.
 
 ```js
-    // If the color picker is too big, try playing with this.
-    // Graphics in tooltips usually look better bigger (especially in HiDPI),
-    // but that can make the tooltips really big. If they are too big,
-    // you can play with this setting.  We compromise with medium.
+    // If the color picker is too big or too small, try playing with this.
+    // This is a coarse control which scales image size relative to the
+    // approximate line height. small (1x), medium (1.5x), and large (2x).
     // (small | medium | large)
     "graphic_size": "medium",
 ```
 
 If you need to set this per OS or per host, you can via [`multiconf`](#multiconf).
+
+### `graphic_scale`
+
+Fine scaling of the size of generated graphics. This overrides `graphic_size`.
+
+```js
+    // Fine scaling of image sizes. Overrides `graphic_size`.
+    // Use an integer or floating point value. `null` disables fine scaling.
+    "graphic_scale": 1.5,
+````
 
 ### `enable_color_picker`
 

